README.md
# DicionarioPT
[This project is not actively maintained anymore, since widgets are no longer supported by macOS.]

## About
*DicionarioPT* is a macOS widget that provides a Portuguese dictionary.
It uses as data source the [Priberam](http://www.priberam.pt/dlpo/).
This widget was originally developed by [José Coelho](http://web.ist.utl.pt/~jose.alberto.coelho/).
Besides the original feature, this widget also provides verbs conjugation.

This widget requires an internet connection.


# Links
* [Downloads](https://github.com/rcgoncalves/dicionariopt/releases/latest)
* [Changelog](https://github.com/rcgoncalves/dicionariopt/blob/master/CHANGELOG.md)


# License
DicionarioPT

Copyright (C) 2006 José Coelho

Copyright (C) 2009-2013 Rui Carlos Gonçalves

This program is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, either version 2 of the License, or (at your option) any later version.

This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with this program.
If not, see <[https://www.gnu.org/licenses/](https://www.gnu.org/licenses/)>.


CHANGELOG.md
# Changelog

## [5.0.1] - 2013-12-09
### Changed
- Update methods to obtain and parse data from Priberam's website.


## [5.0.0] - 2013-09-28
### Changed
- Update methods to obtain and parse data from Priberam's website.
- Improvement the output layout.

### Fixed
- Fix bugs obtaining the suggestions.


## [4.1.0] - 2012-05-09
### Changed
- Improve methods to obtain and parse data from Priberam's website.
- Improvement the output layout.


## [4.0.1] - 2012-04-25
### Changed
- Improvement the output layout.

### Fixed
- Fix bugs in parser.


## [4.0.0] - 2011-09-24
### Changed
- Change the strategies to obtain the data from Priberam's website (it is now used a more reliable method).
- Improvement the output layout.


## [3.3.6] - 2010-03-09
### Fixed
- Fix bug originated by changes in Priberan's website.


## [3.3.5] - 2010-03-07
### Fixed
- Fix bug.


## [3.3.4] - 2010-03-01
### Fixed
- Fix bug originated by changes in Priberan's website.


## [3.3.3] - 2009-10-31
### Fixed
- Fix bug.


## [3.3.2] - 2009-09-21
### Fixed
- Fix bug.


## [3.3.1] - 2009-05-27
### Fixed
- Fix bug originated by changes in Priberan's website.


## [3.3.0] - 2009-05-16
### Added
- Add link to verbs conjugation.
- Add links to words' definitions in verb conjugations.


## [3.2.0] - 2009-05-10
### Changed
- Change method to get definitions.


## [3.1.0] - 2009-05-01
### Changed
- Internal changes and bugs fixes.


## [3.0.0] - 2009-04-21
### Added
- Initial version.


DicionarioPT.dcproj/project/DicionarioPT.wdgt/Parts/parts.js
/* 
 This file was generated by Dashcode and is covered by the 
 license.txt included in the project.  You may edit this file, 
 however it is recommended to first turn off the Dashcode 
 code generator otherwise the changes will be lost. This file
 is for files included by Dashcode directly. It will be replaced
 with an optimized version at deploy time.
*/
var dashcodePartSupport = {
    "scripts": ["/var/folders/qp/qpakv4YQExeXO6JvaasA1k+++TI/-Tmp-/Dashcode/.DicionarioPT.dcprojBtd9br/DicionarioPT.dcproj/project/DicionarioPT.wdgt/Parts/setup.js", "Parts/InfoButton.js"]
};

(function() {
    var scripts = dashcodePartSupport['core'];

    scripts = scripts.concat(dashcodePartSupport['scripts']);

    for(var index in scripts) {
        var path = scripts[index];
        var scriptTag = '<script apple-no-regeneration="yes" type="text/javascript" src="' + path + '"></script>';
        
        document.write( scriptTag );
    }
    
})();



DicionarioPT.dcproj/project/DicionarioPT.wdgt/Scripts/URL.js
/*
 * Copyright (C) 2006, José Coelho (jose.alberto.coelho@gmail.com)
 * This program is free software; you can redistribute it and/or modify it under the terms of the GNU
 * General Public License as published by the Free Software Foundation; either version 2 of the License,
 * or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even
 * the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General
 * Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along with this program; if not,
 * write to the Free Software Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
 *
 */

function URL() {
	this.xmlhttp = new XMLHttpRequest();
	
	// Atributes
	this.location = null;
	this.autoRedirect = true;
	this.outputFile = null;
	this.postData = null;
	this.response = null;
	this.responseData = null;
	this.result = null;
	this.headers = new Object;
	
	_self = this;
	
	// Main Function
	this._getURL = function(url,func,async) {
		var tmpHeaders;
		
		if( url ) {
			_self.location = url;
		}
		
		_self.xmlhttp.open( (_self.postData) ? "POST" : "GET" , _self.location, async);

		_self.xmlhttp.onreadystatechange=function() {
			
			if (_self.xmlhttp.readyState == 4) {
				
				_self.result = _self.xmlhttp.responseText;
				_self.response = _self.xmlhttp.status;
				_self.responseData = _self.xmlhttp.statusText;
				
				tmpHeaders = _self.xmlhttp.getAllResponseHeaders().split("\n");
				for(i = 0; i< tmpHeaders.length; i++) {
					var tmp = tmpHeaders[i].split(":");
					_self.headers[ tmp[0] ] = tmp[1];
				}
				
				if (func) {
					return func(_self)
				} else return _self.result
			}
		}
		
		_self.xmlhttp.setRequestHeader("Cache-Control", "no-cache");
		_self.xmlhttp.send( _self.postData );
	}
}

URL.prototype.fetch = function(url) {
	return this._getURL(url,null,false);
}

URL.prototype.fetchAsync = function(func) {
	return this._getURL(null, func, true);
}

URL.prototype.getResponseHeaders = function(name){
	if(name == "*") {
		return _self.headers;
	}
	else return _self.headers[ name ];
}

URL.prototype.setRequestHeader = function(name, value) {
	_self.xmlhttp.setRequestHeader(name, value);
}

URL.prototype.cancel = function()
{
	_self.xmlhttp.abort();
}

URL.prototype.clear = function()
{
	this.xmlhttp = new XMLHttpRequest();
	this.location = null;
	this.autoRedirect = null;
	this.outputFile = null;
	this.postData = null;
	this.response = null;
	this.responseData = null;
	this.result = null;
	this.headers = new Object;
}


DicionarioPT.dcproj/project/DicionarioPT.wdgt/Scripts/Parser.js
/*
 * Copyright (C) 2013, Rui Carlos Gonçalves (rcgoncalves.pt@gmail.com)
 * This program is free software; you can redistribute it and/or modify it under the terms of the GNU
 * General Public License as published by the Free Software Foundation; either version 2 of the License,
 * or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even
 * the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General
 * Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along with this program; if not,
 * write to the Free Software Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
 */

// ======================================================================================================
// INITIAL PARSING ======================================================================================
// ======================================================================================================

/**
 * Finds the block that contains the relevant content.
 */
function getMainContent() {
  var divs = document.getElementsByTagName('div');
  var div = null;
  var i = divs.length - 1;
  var stop = false;
  while(i >= 0 && !stop) {
    div = divs[i];
    if(div.className != null && div.className.indexOf('pb-main-content') > -1) {
      stop = true;
    }
    i--;
  }
  return div;
}

// ======================================================================================================
// WORD DEFINITION ======================================================================================
// ======================================================================================================

/**
 * Returns a word definition.
 */
function getDefinition(html)
{
  var definition = null;
  if(html != null) {
    if(html.innerHTML.indexOf('alert-info') < 0 ||
        (html.innerHTML.indexOf('Palavra não encontrada') < 0
	&& html.innerHTML.indexOf('Palavra reconhecida pelo FLiP mas sem definição no dicionário') < 0)
	) {
      html = html.getElementsByTagName('div')[0];
      html = parseDefinitionStructure(html);
      if(html != null) {
        definition = html.innerHTML;
	definition = stripTags(definition, true,
	    'font',
	    'img',
	    'notas', '/notas',
	    'categoria_ext_aao', '/categoria_ext_aao',
	    'pt', '/pt',
	    'aao', '/aao',
	    'dominio_ext_aao', '/dominio_ext_aao');
      }
    }
  }
  return definition;
}

/**
 * Parses a word definition.
 * Adjust URLs, removes some attributes, remove some tags, etc.
 */
function parseDefinitionStructure(html) {
  for(var i = html.children.length ; i > 0; i--) {
    var elem = html.children[i-1];
    var toDel = 0;
    switch(elem.tagName) {
      case 'A' :
        var url=elem.getAttribute('href');
        if(url!=null) {
          if(!url.match(/\/dlpo\//ig)) {
            elem.removeAttribute('href');
          }
          else {
            url=url.replace(/(.|\n)*\/dlpo\/Conjugar\/(.+)/ig, "javascript:Conjugation('$2')");
	    url=url.replace(/(.|\n)*\/dlpo\/(.+)/ig, "javascript:Search('$2')");
            elem.setAttribute('href', url);
          }
        }
        break;
      case 'REFS_EXTERNAS' :
      case 'SCRIPT' :
        toDel = 1;
        break;
      case 'DIV' :
        var style = elem.style;
	if(style.display == 'none') toDel = 1;
        break;
      case 'TD' :
      case 'TABLE' :
        removeAttributes(elem);
        break;
      case 'SPAN' :
        if(elem.className == 'varpb') toDel = 1;
	if(elem.className == 'dAO') toDel = 1;
        break;
      case 'BR' :
        toDel = 2;
	break;
      case 'CATEGORIA_EXT_AAO' :
        var newElem = document.createElement('SPAN');
	newElem.className = 'dpt-categoria';
	newElem.innerHTML = elem.innerHTML;
	elem.innerHTML = '';
	elem.appendChild(newElem);
        break;
    }
    if(toDel == 1) {
      html.removeChild(elem);
      if(i > 1 && html.children[i-2].tagName == 'BR') {
        html.removeChild(html.children[i-2]);
        i--;
      }
    }
    else if(toDel == 2 && i > 1 && html.children[i-2].tagName == 'BR') {
      html.removeChild(elem);
    }
    else if(elem.innerHTML.match(/^\s*$/) && toDel != 2) {
      html.removeChild(elem);
    }
    else {
      elem.removeAttribute('style');
      // recursive call
      parseDefinitionStructure(elem);
    }
  }
  return html;
}

// ======================================================================================================
// SUGGESTIONS ==========================================================================================
// ======================================================================================================

/**
 * Returns a feedback message for words that do not have a definition available.
 */
function getSuggestions(html) {
  var result = "";
  if(html.innerHTML.indexOf('Palavra reconhecida pelo FLiP mas sem definição no dicionário') > -1) {
    result = '<p>Palavra reconhecida mas sem definição no disponível.</p>';
  }
  else {
    var list = getSuggestionsList(html);
    for(i in list) {
      result += '<li><a href="javascript:Search(\'' + list[i] + '\')">' + list[i] + '</a></li>\n';
    }
    if(result != '') {
      result = '<p style="font-weight:bold;">Palavra não encontrada.</strong></p><p>Experimente uma das seguintes sugestões:</p>\n'
          + '<ul style="margin: 3px; padding-left: 18px; text-indent: -3px; list-style-type: square;">\n'
	  + result
	    + '</ul>\n';
    }
    else {
      result = '<p style="font-weight:bold;">Palavra não encontrada.</p><p>Não existem sugestões.</p>';
    }
  }
  return result;
}

/**
 * Returns list of suggestions.
 */
function getSuggestionsList(html) {
  var result = new Array();
  var divs = html.getElementsByTagName('div');
  for(var i = 0; i < divs.length; i++) {
    var div = divs[i];
    if(div.id == 'FormataSugestoesENaoEncontrados') {
      result.push(div.innerText);
    }
  }
  return result;
}

// ======================================================================================================
// CONJUGATION ==========================================================================================
// ======================================================================================================

/**
 * Returns a verb conjugation.
 */
function getConjugation(html) {
  var result = html.children[0];
  // remove meta tags
  var elems = result.getElementsByTagName('meta');
  for(var i = elems.length - 1; i >= 0; i--) {
    var elem = elems[i];
    elem.parentNode.removeChild(elem);
  }
  // remove style tags
  elems = result.getElementsByTagName('style');
  for(var i = elems.length - 1; i >= 0; i--) {
    var elem = elems[i];
    elem.parentNode.removeChild(elem);
  }
  // remove span tags
  elems = result.getElementsByTagName('span');
  for(var i = elems.length - 1; i >= 0; i--) {
    var elem = elems[i];
    if(elem.style.color == 'rgb(255, 255, 255)') {
      elem.parentNode.removeChild(elem);
    }
  }
  var regex = new RegExp('/ elas?/ vocês?', 'g');
  return result.innerHTML.replace(regex, '');
}

// ======================================================================================================
// AUXILIAR FUNCTIONS ===================================================================================
// ======================================================================================================

/**
 * Remove the attributes of a HTMLElement.
 */
function removeAttributes(element) {
  while(element.attributes.length > 0) {
    element.removeAttribute(element.attributes[0].name);
  }
}

DicionarioPT.dcproj/project/DicionarioPT.wdgt/Scripts/Widget.js
/*
 * Copyright (C) 2006, José Coelho (jose.alberto.coelho@gmail.com)
 * This program is free software; you can redistribute it and/or modify it under the terms of the GNU
 * General Public License as published by the Free Software Foundation; either version 2 of the License,
 * or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even
 * the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General
 * Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along with this program; if not,
 * write to the Free Software Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
 *
 */


var Widget = new Object();

//The closed size for the widget
Widget.CLOSED_SIZE = { x: 450, y: 75 };

//The minimum size for the widget (when open)
Widget.MINIMUM_SIZE = { x: 450, y: 250 };

//The last user resize for the widget
Widget.LAST_SIZE = { x: Widget.MINIMUM_SIZE.x, y: Widget.MINIMUM_SIZE.y }

Widget.close = false;

Widget.setSize = function() {
	document.getElementById("front").style.width = Widget.CLOSED_SIZE.x + 'px';
	document.getElementById("front").style.height = Widget.CLOSED_SIZE.y + 'px';
}

Widget.shrink = function() {
	Widget.close = true;
	document.getElementById("bottomBar").style.display = "none";
	
	Widget.LAST_SIZE.x = parseInt(document.getElementById("front").style.width);
	Widget.LAST_SIZE.y = parseInt(document.getElementById("front").style.height);
	
	Widget.resize( parseInt(document.getElementById("front").style.width), Widget.CLOSED_SIZE.y );

}

Widget.expand = function() {
	Widget.close = false;
	document.getElementById("bottomBar").style.display = "block";
	
	if( Widget.LAST_SIZE.y > Widget.MINIMUM_SIZE.y && Widget.LAST_SIZE.x > Widget.MINIMUM_SIZE.x )
		Widget.resize( Widget.LAST_SIZE.x , Widget.LAST_SIZE.y );
	else
		Widget.resize( Widget.MINIMUM_SIZE.x , Widget.MINIMUM_SIZE.y );
}

Widget.needResize = function() {
	return Widget.close;
}

Widget.eventInset = null;
Widget.resizeMouseDown = function(event) {
	document.addEventListener("mousemove", Widget.resizeMouseMove, true);
	document.addEventListener("mouseup", Widget.resizeMouseUp, true);
	
	Widget.eventInset = {x:(window.innerWidth - event.x), y:(window.innerHeight - event.y)};

	event.stopPropagation();
	event.preventDefault();
}

Widget.resizeMouseMove = function(event) {
	var x = event.x + Widget.eventInset.x;
	var y = event.y + Widget.eventInset.y;
	
	if (x < Widget.MINIMUM_SIZE.x) {
		x = Widget.MINIMUM_SIZE.x;
	}

	if (y < Widget.MINIMUM_SIZE.y) {
		y = Widget.MINIMUM_SIZE.y;
	}

	Widget.resize(x,y);
	
	scrollArea.refresh();
	event.stopPropagation();
	event.preventDefault();
}

Widget.resizeMouseUp = function(event) {	
	document.removeEventListener("mousemove", Widget.resizeMouseMove, true);
	document.removeEventListener("mouseup", Widget.resizeMouseUp, true);

	event.stopPropagation();
	event.preventDefault();
}

// Resize the widget.
Widget.resize = function(x, y) {
	document.getElementById("front").style.height = y + 'px';
	document.getElementById("front").style.width = x + 'px';
	window.resizeTo(x, y);
}

var progress = null;
Widget.progressLoad = function(element) {
	progress = new ProgressIndicator(element);
}

Widget.progressStart = function() {
	progress.start();
}

Widget.progressStop = function() {
	progress.stop();
}

function ProgressIndicator(element) {
    this.count = 0;
    this.timer = null;
    this.element = document.getElementById(element);
    this.element.style.display = "none";
    this.imageBaseURL = "Images/Progress/prog";
}

ProgressIndicator.prototype = {
    start : function () {
        this.element.style.display = "block";        
        if (this.timer) clearInterval(this.timer);
        this.tick();
        var localThis = this;
        this.timer = setInterval (function() { localThis.tick() }, 60);
    },

    stop : function () {
        clearInterval(this.timer);
        this.element.style.display = "none";
    },

    tick : function () {
        var imageURL = this.imageBaseURL + (this.count + 1) + ".png";
        this.element.src = imageURL;
        this.count = (this.count + 1) % 12;
    }
}



DicionarioPT.dcproj/project/DicionarioPT.wdgt/Scripts/Dicionario.js
/*
 * Copyright (C) 2006, José Coelho (jose.alberto.coelho@gmail.com)
 * Copyright (C) 2013, Rui Carlos Gonçalves (rcgoncalves.pt@gmail.com)
 * This program is free software; you can redistribute it and/or modify it under the terms of the GNU
 * General Public License as published by the Free Software Foundation; either version 2 of the License,
 * or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even
 * the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General
 * Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along with this program; if not,
 * write to the Free Software Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
 *
 */

var Web;
var scrollArea;
var scrollBar;

function LoadWidget() {
  Web = new URL();
  Widget.setSize();
  Widget.progressLoad("progressGraphic");
  scrollBar = new AppleVerticalScrollbar(document.getElementById('scrollBar'));
  scrollArea = new AppleScrollArea(document.getElementById('scrollArea'), scrollBar);
  scrollArea.addScrollbar(scrollBar);
  scrollBar.show();
  Widget.shrink();
}

function shrinkExpand(event) {
  if(Widget.close) {
    Widget.expand();
  }
  else {
    Widget.shrink();
  }
}

/**
 * Searches for a word definition.
 */
function Search(value) {
  document.getElementById("SearchField").value = value;
  if(value.length == 0) {
    Widget.shrink();
    document.getElementById("contents").innerHTML = '';
  }
  else {
    Widget.progressStart();
    Web.location = "http://www.priberam.pt/dlpo/" + value;
    Web.fetchAsync(GetDefinition);
  }
}

/**
 * Callback function that processes a word definition.
 */
function GetDefinition(webfetch) {
  document.getElementById('contents').innerHTML = webfetch.result;
  var content = getMainContent();
  document.getElementById('contents').innerHTML = '';
  var def = getDefinition(content);
  if(def) {
    document.getElementById('contents').innerHTML = def;
  }
  else {
    document.getElementById('contents').innerHTML = getSuggestions(content);
  }
  if(Widget.needResize()) {
    Widget.expand();
  }
  Widget.progressStop();
  scrollArea.refresh();
  scrollBar.refresh();
  scrollArea.verticalScrollTo(0);
}

/**
 * Obtains a verb conjugation.
 */
function Conjugation(value) {
  document.getElementById('SearchField').value = value;
  if( value.length == 0 ) {
    Widget.shrink();
    document.getElementById('contents').innerHTML = '';
  }
  else {
    Widget.progressStart();
    Web.location = 'http://www.priberam.pt/dlpo/Conjugar/' + value;
    Web.fetchAsync( GetConjugation );
  }
}

/**
 * Callback function that processes a verb conjugation.
 */
function GetConjugation(webfetch) {

  document.getElementById('contents').innerHTML = webfetch.result;
  var content = getMainContent(webfetch.result);
  document.getElementById('contents').innerHTML = '';
  document.getElementById('contents').innerHTML = getConjugation(content);
  if(Widget.needResize()) {
    Widget.expand();
  }
  Widget.progressStop();
  scrollArea.refresh();
  scrollBar.refresh();
  scrollArea.verticalScrollTo(0);
}

/**
 * Opens widget website.
 */
function openWebsite()  {
  widget.openURL('http://rcgoncalves.pt/project/dicionariopt/');
}

/*
 * Removes html tags or make them a "clean" version
 * first argument - the string to clean
 * second argument - bool - dummy tag or remove
 * other arguments - html tags to clean/remove
 */
function stripTags() {
  var string = arguments[0];
  var strip = arguments[1];
  for(var i = 2; i < arguments.length; i++)
  {
    var regex = new RegExp('<' + arguments[i] + '[^>]*>', 'img');
    string = string.replace(regex, strip ? '' : '<' + arguments[i] + '>');
  }
  return string;
}